lib/puppet/functions/type_of.rb

```diff
@@ -0,0 +1,13 @@
+# Returns the type when passed a value.
+#
+# @example how to compare values' types
+#   # compare the types of two values
+#   if type_of($first_value) != type_of($second_value) { fail("first_value and second_value are different types") }
+#
```

Comment:
Can add this as well:

@example how to compare against a type
  unless type_of($first_value) <= Numeric { fail("first_value must be Numeric") } 
  unless type_of{$first_value) <= Collection[1] { fail("first_value must be an Array or Hash, and contain at least one element") }